src/package-transpilation-registry.js

```diff
@@ -0,0 +1,151 @@
+var crypto = require('crypto')
+var fs = require('fs')
+var path = require('path')
+
+var minimatch = require('minimatch')
+
+var Resolve = null
+
+function PackageTranspilationRegistry () {
+  this.configByPackagePath = {}
+  this.specByFilePath = {}
+  this.transpilerPaths = {}
+}
+
+Object.assign(PackageTranspilationRegistry.prototype, {
+  addTranspilerConfigForPath: function (packagePath, config) {
+    this.configByPackagePath[packagePath] = {
+      specs: config,
+      path: packagePath
+    }
+  },
+
+  removeTranspilerConfigForPath: function (packagePath) {
+    delete this.configByPackagePath[packagePath]
+  },
+
+  // Wraps the transpiler in an object with the same interface
+  // that falls back to the original transpiler implementation if and
+  // only if a package hasn't registered its desire to transpile its own source.
+  wrapTranspiler: function (transpiler) {
+    var self = this
+    return {
+      getCachePath: function (sourceCode, filePath) {
+        var spec = self.getPackageTranspilerSpecForFilePath(filePath)
+        if (spec) {
+          return self.getCachePath(sourceCode, filePath, spec)
+        }
+
+        return transpiler.getCachePath(sourceCode, filePath)
+      },
+
+      compile: function (sourceCode, filePath) {
+        var spec = self.getPackageTranspilerSpecForFilePath(filePath)
+        if (spec) {
+          return self.transpileWithPackageTranspiler(sourceCode, filePath, spec)
+        }
+
+        return transpiler.compile(sourceCode, filePath)
+      },
+
+      shouldCompile: function (sourceCode, filePath) {
+        if (self.transpilerPaths[filePath]) {
+          return false
+        }
+        var spec = self.getPackageTranspilerSpecForFilePath(filePath)
+        if (spec) {
+          return true
+        }
+
+        return transpiler.shouldCompile(sourceCode, filePath)
+      }
+    }
+  },
+
+  getPackageTranspilerSpecForFilePath: function (filePath) {
+    if (this.specByFilePath[filePath] !== undefined) return this.specByFilePath[filePath]
+
+    var config = null
+    var spec = null
+    var thisPath = filePath
+    var lastPath = null
+    // Iterate parents from the file path to the root, checking at each level
+    // to see if a package manages transpilation for that directory.
+    // This means searching for a config for `/path/to/file/here.js` only
+    // only iterates four times, even if there are hundreds of configs registered.
+    while (thisPath !== lastPath) { // until we reach the root
+      if (config = this.configByPackagePath[thisPath]) { // eslint-disable-line no-cond-assign
+        for (var i = 0; i < config.specs.length; i++) {
+          spec = config.specs[i]
+          if (minimatch(filePath, path.join(config.path, spec.glob))) {
```

Comment:
Maybe you want to compile the minimatch regexp up front? Not sure if they do caching internally.